lib/elixir/lib/regex.ex

```diff
@@ -397,30 +404,72 @@ defmodule Regex do
   defp parts_to_index(:infinity),                      do: 0
   defp parts_to_index(n) when is_integer(n) and n > 0, do: n
 
-  defp do_split(_, string, offset, _counter, true) when byte_size(string) <= offset,
+  defp do_split(_, string, offset, _counter, true, _with_captures) when byte_size(string) <= offset,
     do: []
 
-  defp do_split(_, string, offset, 1, _trim),
+  defp do_split(_, string, offset, 1, _trim, _with_captures),
     do: [binary_part(string, offset, byte_size(string) - offset)]
 
-  defp do_split([], string, offset, _counter, _trim),
+  defp do_split([], string, offset, _counter, _trim, _with_captures),
     do: [binary_part(string, offset, byte_size(string) - offset)]
 
-  defp do_split([[{pos, _}|h]|t], string, offset, counter, trim) when pos - offset < 0,
-    do: do_split([h|t], string, offset, counter, trim)
+  defp do_split([[{pos, _}|h]|t], string, offset, counter, trim, with_captures) when pos - offset < 0,
+    do: do_split([h|t], string, offset, counter, trim, with_captures)
 
-  defp do_split([[]|t], string, offset, counter, trim),
-    do: do_split(t, string, offset, counter, trim)
+  defp do_split([[]|t], string, offset, counter, trim, with_captures),
+    do: do_split(t, string, offset, counter, trim, with_captures)
+
+  defp do_split([[{pos, length}|h]|t], string, offset, counter, true, false) do
+    new_offset = pos + length
+    keep = pos - offset
+
+    if keep == 0 do
+      do_split([h|t], string, new_offset, counter, true, false)
+    else
+      <<_::binary-size(offset), part::binary-size(keep), _::binary>> = string
+      [part | do_split([h|t], string, new_offset, counter - 1, true, false)]
+    end
+  end
 
-  defp do_split([[{pos, length}|h]|t], string, offset, counter, trim) do
+  defp do_split([[{pos, length}|h]|t], string, offset, counter, false, false) do
     new_offset = pos + length
     keep = pos - offset
 
-    if keep == 0 and (length == 0 or trim) do
-      do_split([h|t], string, new_offset, counter, trim)
+    if keep == 0 and length == 0 do
+      do_split([h|t], string, new_offset, counter, false, false)
     else
       <<_::binary-size(offset), part::binary-size(keep), _::binary>> = string
-      [part|do_split([h|t], string, new_offset, counter - 1, trim)]
+      [part | do_split([h|t], string, new_offset, counter - 1, false, false)]
+    end
+  end
+
+  defp do_split([[{pos, length}|h]|t], string, offset, counter, true, true) do
```

Comment:
I had originally thought to split only in terms of include_captures, so it would be in 2 clauses and not 4. :) This way only the first 2 lines of each clause would be the same (he ones where we calculate both new_offset and keep). Everything else would be different.